source/FreeRTOS_IP_Multicast.c

```diff
@@ -0,0 +1,983 @@
+         pxIterator != ( const ListItem_t * ) xEnd;
+         pxIterator = ( const ListItem_t * ) listGET_NEXT( pxIterator ) )
+    {
+        pxMRD = ( MCastReportData_t * ) listGET_LIST_ITEM_OWNER( pxIterator );
+
+        if( ( pxInterface == pxMRD->pxInterface ) || ( pxMRD->pxInterface == NULL ) )
+        {
+            if( ( pxMRD->xCountDown < 0 ) || ( pxMRD->xCountDown >= xMaxCountdown ) )
+            {
+                if( xApplicationGetRandomNumber( &( ulRandom ) ) == pdFALSE )
+                {
+                    /* The random number generator failed. Schedule the report for immediate sending. */
+                    ulRandom = 0;
+                }
+                else
+                {
+                    /* Everything went well. ulRandom now holds a good random value. */
+                }
+
+                /* Generate a random countdown between 0 and usMaxCountdown - 1. */
+                pxMRD->xCountDown = ulRandom % xMaxCountdown;
+            }
+            else
+            {
+                /* The report is already scheduled for sooner than xMaxCountdown. Do nothing. */
+            }
+        }
+        else
+        {
+            /* The multicast report is for a different interface. Do nothing. */
+        }
+    } /* for(;;) iterating over prvMulticastReportsList */
+}
+
+void prvScheduleMulticastReports( BaseType_t xIsIPv6,
+                                  void * pvGroupAddress,
+                                  NetworkInterface_t * pxInterface,
+                                  BaseType_t xMaxCountdown )
+{
+    const ListItem_t * pxIterator;
+    const ListItem_t * xEnd;
+    uint32_t ulRandom;
+    MCastReportData_t * pxMRD;
+    const IPv6_Address_t * pxIPv6_GroupAddress;
+    static uint32_t ulNonRandomCounter = 0; /* In case the random number generator fails have a "not so random number" ready. */
+
+    /* Go through the list of IGMP reports and schedule them. Note, the IGMP event is set at 100ms */
+
+    /* Sanity enforcement. Technically, there is nothing wrong with trying to schedule reports "right away",
+     * but having xMaxCountdown = 0 will cause problems with the modulo operation further down. */
+    xMaxCountdown = max( 1, xMaxCountdown );
+
+    xEnd = listGET_END_MARKER( &prvMulticastReportsList );
+
+    for( pxIterator = ( const ListItem_t * ) listGET_NEXT( xEnd );
+         pxIterator != ( const ListItem_t * ) xEnd;
+         pxIterator = ( const ListItem_t * ) listGET_NEXT( pxIterator ) )
+    {
+        BaseType_t xReschedule = pdFALSE;
+        pxMRD = ( MCastReportData_t * ) listGET_LIST_ITEM_OWNER( pxIterator );
+
+        /* Make sure IP versions are the same. */
+        if( xIsIPv6 != pxMRD->xMCastGroupAddress.xIs_IPv6 )
+        {
+            continue;
+        }
+
+        if( xIsIPv6 )
+        {
+            /* IPv6 MLD */
+            pxIPv6_GroupAddress = ( const IPv6_Address_t * ) pvGroupAddress;
+            BaseType_t i, xSpecificQuery = pdFALSE;
```

Comment:
`i` should be unsigned as `ipSIZE_OF_IPv6_ADDRESS` is unsigned.

Reply:
Fixed by completely redoing the specific query check. Back when I wrote this I did not know FreeRTOS_in6addr_any existed.